cisc/certificate.go

```diff
@@ -0,0 +1,321 @@
+	if reErr != nil {
+		log.Fatal("error while revoke:", reErr)
+	}
+	log.Print("Succesfully revoked")
+}
+
+//code addapted from :https://ericchiang.github.io/post/go-letsencrypt/ "
+func getCert(dir string, domain string) string {
+	// Create a client to the acme of letsencrypt.
+	log.Info("Requesting RSA keys")
+	cli, err := letsencrypt.NewClient("https://acme-v01.api.letsencrypt.org/directory")
+	//cli, err := letsencrypt.NewClient("https://acme-staging.api.letsencrypt.org/directory")
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if _, err = os.Stat("/home/" + dir + "/cert/" + domain); os.IsNotExist(err) {
+		os.MkdirAll("/home/"+dir+"/cert/"+domain, 0777)
+	}
+
+	// Find or generate the private key to register on the ACME server.
+	var key *rsa.PrivateKey
+	key, err = loadKey("/home/" + dir + "/cert/" + domain + "/registerkey.pem")
+	if err != nil {
+		log.Print("new generate RSA key for register , keep it safe")
+		key, err = rsa.GenerateKey(rand.Reader, 2048)
+		if err != nil {
+			log.Fatal(err)
+		}
+		ioutil.WriteFile("/home/"+dir+"/cert/"+domain+"/registerkey.pem", pem.EncodeToMemory(&pem.Block{Bytes: x509.MarshalPKCS1PrivateKey(key), Type: "RSA PRIVATE KEY"}), 0644)
+	}
+
+	//register to the ACME server
+	if _, err := cli.NewRegistration(key); err != nil {
+		log.Fatal(err)
+	}
+
+	log.Print("Registration successful!")
+
+	//Get the http-challenge
+	auth, _, err := cli.NewAuthorization(key, "dns", domain)
+	if err != nil {
+		log.Fatal(err)
+	}
+	httpChalURL := ""
+	for _, chal := range auth.Challenges {
+		if chal.Type == "http-01" {
+			httpChalURL = chal.URI
+		}
+	}
+	if httpChalURL == "" {
+		log.Fatal("Not found http-01 challenge")
+	}
+	chal, err := cli.Challenge(httpChalURL)
+	if err != nil {
+		log.Fatal(err)
+	}
+	//Get the path and the resource needed to perform the challenge
+	path, resource, err := chal.HTTP(key)
+	if err != nil {
+		log.Fatal(err)
+	}
+	//copy the resource for the challenge
+	if os.MkdirAll("/home/"+dir+"/www/.well-known/acme-challenge/", 0711) != nil {
+		log.Fatal("problem creating new dir for challenge :", err)
+	}
+	if err := ioutil.WriteFile("/home/"+dir+"/www/"+path, []byte(resource), 0644); err != nil {
+		log.Fatal(err)
+	}
+	//complete the challenge
+	if err := cli.ChallengeReady(key, chal); err != nil {
+		log.Fatal("Challenge not complete :", err)
+	}
+	log.Print("Challenge Complete")
+	//create a certificate request
+	csr, _, err := newCSR(domain, dir)
+	if err != nil {
+		log.Fatal(err)
+	}
+	// Request a certificate for the domain
+	cert, err := cli.NewCertificate(key, csr)
+
+	if err != nil {
+		log.Fatal(err)
+	}
+	log.Print("New Certificate Successfuly created")
+	fullchain, err := cli.Bundle(cert)
+
+	if err != nil {
+		log.Fatal("can't bundles the certificate:", err)
+	}
+	ioutil.WriteFile("/home/"+dir+"/cert/"+domain+"/fullchain.pem", fullchain, 0644)
+
+	return string(fullchain)
+}
+
+//code adapted from :https://ericchiang.github.io/post/go-letsencrypt/ "
```

Comment:
Better comment - keep where it comes from, but also explain what it's doing here.